File: Assignment 4/4-3.py
"""
    과제 4-(3) 정렬

    내장된 정렬 방법과 비교하기
"""

from sortings import sorter
import time

import numpy as np
from numba import jit

if __name__ == "__main__":
    # 1. Python 기본 제공 정렬
    s = sorter()
    start = time.time()
    s.array.sort(key=lambda x: x[0]) # key를 중심으로 sort
    builtin = time.time() - start
    s.check_sorted()
    
    # 2. 새로 만든 퀵 정렬
    s = sorter()
    start = time.time()
    s.quick()
    my = time.time() - start
    s.check_sorted()

    # 3. 외부 라이브러리(NumPy) 사용
    s = sorter()
    s.array = np.array(s.array, dtype=[('key', 'U30'), ('value', 'U50')])
    start = time.time()
    s.array.sort(order='key')
    numpy_time = time.time() - start
    s.check_sorted()

    print(f"Python 기본 제공 정렬: {builtin:.3f}초")
    print(f"새로 만든 퀵 정렬: {my:.3f}초")
    print(f"NumPy 정렬: {numpy_time:.3f}초")

File: Assignment 4/4-3.c
/**
 * 과제 4-(3) 정렬
 * 
 * 내장된 정렬 방법과 비교하기
*/

#include "sortings.h"

int compare(const void *a, const void *b) {
    return strcmp(((record *)a)->key, ((record *)b)->key);
}

int main(void) {
    // C에서는 quick sort를 기본으로 제공한다.
    // qsort()와 내가 만든 quick_sort를 비교한다.
    record *array1 = (record *)calloc(WORD_COUNT, sizeof(record));
    record *array2 = (record *)calloc(WORD_COUNT, sizeof(record));

    load(array1);
    load(array2);

    // 1. 내장 qsort() 사용
    clock_t start = clock();
    qsort(array1, WORD_COUNT, sizeof(record), compare);
    clock_t builtin = clock() - start;
    check_sorted(array1);

    // 2. 새로 만든 quick_sort() 사용
    start = clock();
    quick_sort(array2);
    clock_t my = clock() - start;
    check_sorted(array2);

    printf("C에서 내장된 정렬 방법(qsort): %.3f초\n", (double)builtin / CLOCKS_PER_SEC);
    printf("새로 만든 정렬 방법: %.3f초\n", (double)my / CLOCKS_PER_SEC);

    free(array1);
    free(array2);
    
    return 0;
}